### PR TITLE
osd/OSD: rewrite send_incremental_map()

### PR DESCRIPTION
**Pull request by Matan-B**
* See: 9fba69a11aa940ed36339bb24b05cb92165db516
   build_incremental_map_msg() no longer returns NULL on failures.
   while(!m) if-case is removed.

* Send all relevant maps even when adjusting `since`

* Simplify code readability

* Fix log message

* Avoid bufferlist copying

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "quincy"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

- When filling out the below checklist, you may click boxes directly in the GitHub web UI.  When entering or editing the entire PR message in the GitHub web UI editor, you may also select a checklist item by adding an `x` between the brackets: `[x]`.  Spaces and capitalization matter when checking off items this way.

## Checklist
- Tracker (select at least one)
  - [ ] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [x] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [x] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
- `jenkins test rook e2e`
</details>

### CODE REVIEW

File: src/osd/OSD.cc [hunk 1]
```diff
@@ -1457,22 +1457,19 @@ void OSDService::send_incremental_map(epoch_t since, Connection *con,
            << " to " << con << " " << con->get_peer_addr() << dendl;
 
   MOSDMap *m = NULL;
```
Comment (ronen-fr) [inline src/osd/OSD.cc:1459]: you should probably avoid initializing 'm' (and surely not to null).

Comment (ronen-fr) [reply src/osd/OSD.cc:1459]: Other than that - looks good

Comment (athanatos) [reply src/osd/OSD.cc:1459]: @ronen-fr Why do you want this to be uninitialized?  Ideally, we'd simply initialize it to the correct value, but failing that I really don't want it left uninitialized -- if the below if/else is incorrectly modified to fail to set it it a null value will be far easier to debug.

Comment (athanatos) [reply src/osd/OSD.cc:1459]: Converting the if/else to a lambda that returns m into send_map would, of course, eliminate the problem.

File: src/osd/OSD.cc [hunk 2]
```diff
@@ -1457,22 +1457,19 @@ void OSDService::send_incremental_map(epoch_t since, Connection *con,
...
-			       osdmap->get_encoding_features());
-      m->cluster_osdmap_trim_lower_bound = sblock.cluster_osdmap_trim_lower_bound;
-      m->newest_map = sblock.get_newest_map();
-      get_map_bl(to, m->maps[to]);
-      send_map(m, con);
-      return;
-    }
-
+  OSDSuperblock sblock(get_superblock());
+  if (since < sblock.get_oldest_map()) {
```
Comment (athanatos) [inline src/osd/OSD.cc:1461]: I think this patch PR is correct as is.

Going a bit further, I'm not actually sure why this if branch works this way.  build_incremental_map_msg handles the case (since < oldest_map) by sending oldest_map and then incrementals up to the current map.  That strikes me as probably the right behavior, so it might be that we can actually drop this if entirely.  I might be missing something though.

**Review by athanatos — APPROVED** [review]

File: src/osd/OSD.cc (hunk 2 above)
Comment (Matan-B) [reply src/osd/OSD.cc]: > I think this patch is correct as is.
> 
> Going a bit further, I'm not actually sure why this if branch works this way. build_incremental_map_msg handles the case (since < oldest_map) by sending oldest_map and then incrementals up to the current map. That strikes me as probably the right behavior, so it might be that we can actually drop this if entirely. I might be missing something though.

Moreover, build_incremental_map_msg already adjusts `since` according to the trim_lower_bound epoch. Pushed cb675b13748c4d4588c38dee46eed114c0165201

**Comment by Matan-B** [comment]
Found few more cleanups/optimizations. Re-requesting review since it's no longer a cleanup.

File: src/osd/OSD.cc [hunk 3]
```diff
@@ -1388,21 +1388,13 @@ MOSDMap *OSDService::build_incremental_map_msg(epoch_t since, epoch_t to,
   if (since < m->cluster_osdmap_trim_lower_bound) {
```
Comment (ronen-fr) [inline src/osd/OSD.cc:1388]: is 'max' used at all now?

File: src/osd/OSD.cc [hunk 4]
```diff
@@ -1444,41 +1436,23 @@ MOSDMap *OSDService::build_incremental_map_msg(epoch_t since, epoch_t to,
...
 
-    if (to > since && (int64_t)(to - since) > cct->_conf->osd_map_share_max_epochs) {
-      dout(10) << "  " << (to - since) << " > max " << cct->_conf->osd_map_share_max_epochs
-	       << ", only sending most recent" << dendl;
-      since = to - cct->_conf->osd_map_share_max_epochs;
-    }
-
-    m = build_incremental_map_msg(since, to, sblock);
+  OSDSuperblock sblock(get_superblock());
+  if (to > since && (int64_t)(to - since) > cct->_conf->osd_map_share_max_epochs) {
```
Comment (ronen-fr) [inline src/osd/OSD.cc:1449]: to-since is a problem by itself. I'd suggest verifying that 'to' > 'since' before any 'to - since'

File: src/osd/OSD.cc [hunk 5]
```diff
@@ -7376,8 +7350,9 @@ void OSDService::maybe_share_map(
                << session->projected_epoch << dendl;
       return;
     }
-
-    send_from = session->projected_epoch;
+    // send incremental maps in the range of:
+    // (projected_epoch, osdmap]
+    send_from = session->projected_epoch + 1;
```
Comment (ronen-fr) [inline src/osd/OSD.cc:7355]: (not really part of this PR, but as I needed a moment to get over the indentation: )
I'd rather replace the block with 'unique_lock unlocked before the send_..

File: src/osd/OSD.cc (hunk 4 above)
Comment (Matan-B) [reply src/osd/OSD.cc:1449]: It's verified:
`if (to > since &&`

File: src/osd/OSD.cc (hunk 3 above)
Comment (Matan-B) [reply src/osd/OSD.cc:1388]: Yes

File: src/osd/OSD.cc (hunk 5 above)
Comment (Matan-B) [reply src/osd/OSD.cc:7355]: Ack, can be addressed outside the scope of this PR.

File: src/osd/OSD.cc (hunk 4 above)
Comment (ronen-fr) [reply src/osd/OSD.cc:1449]: missed that...

File: src/osd/OSD.cc (hunk 3 above)
Comment (ronen-fr) [reply src/osd/OSD.cc:1388]: Probably missed something, but I did not see it used in the code

File: src/osd/OSD.cc [hunk 6]
```diff
@@ -1388,21 +1388,13 @@ MOSDMap *OSDService::build_incremental_map_msg(epoch_t since, epoch_t to,
...
              << " > since " << since << ", starting with full map"
              << dendl;
     since = m->cluster_osdmap_trim_lower_bound;
-    if (!get_map_bl(since, bl)) {
-      derr << __func__ << " missing full map " << since << dendl;
-      goto panic;
-    }
-    max--;
-    max_bytes -= bl.length();
-    m->maps[since] = std::move(bl);
```
Comment (athanatos) [inline src/osd/OSD.cc:1403]: This removal is wrong.  The below loop populates the map message with incremental maps.  This branch specifically handles the case where the receiving OSD will have a map gap.  In that case, the first map we send must be a full map.

File: src/osd/OSD.cc [hunk 7]
```diff
@@ -1374,6 +1374,18 @@ void OSDService::got_stop_ack()
...
+int& max,
+ssize_t& max_bytes,
+bufferlist& bl,
+std::function<bool()> map_getter) {
+  if (!map_getter()) {
+    throw -ENOENT;
+  }
+  max--;
+  max_bytes -= bl.length();
+}
```
Comment (athanatos) [inline src/osd/OSD.cc:1387]: This should be a lambda local to build_incremental_map_message -- every caller passes the same map_getter.

Just return ENOENT, please don't introduce exceptions unnecessarily.

Comment (athanatos) [reply src/osd/OSD.cc:1387]: This should probably return std::optional<bufferlist>

Comment (athanatos) [reply src/osd/OSD.cc:1387]: Probably rename to get_map_and_adjust_counters

Comment (Matan-B) [reply src/osd/OSD.cc]: Renamed, made local and found it easier to return a boolean instead.

**Review by athanatos — APPROVED** [review]

**Comment by Matan-B** [comment]
jenkins test make check

**Comment by Matan-B** [comment]
jenkins test api

**Comment by ljflores** [comment]
@Matan-B there is a crash showing up in teuthology tests that seems related:

https://pulpito.ceph.com/yuriw-2023-12-12_16:18:21-rados-wip-yuri-testing-2023-12-11-1524-distro-default-smithi/7488633/
```
2023-12-12T16:52:48.574 INFO:tasks.ceph.osd.0.smithi141.stderr:/home/jenkins-build/build/workspace/ceph-dev-new-build/ARCH/x86_64/AVAILABLE_ARCH/x86_64/AVAILABLE_DIST/centos9/DIST/centos9/MACHINE_SIZE/gigantic/release/19.0.0-126-gecf1bb4f/rpm/el9/BUILD/ceph-19.0.0-126-gecf1bb4f/src/osd/OSDMap.cc: 2266: FAILED ceph_assert(inc.epoch == epoch+1)
2023-12-12T16:52:48.576 INFO:tasks.ceph.osd.0.smithi141.stderr: ceph version 19.0.0-126-gecf1bb4f (ecf1bb4fb196fceec77b1f6d38d55ac4991874ac) reef (dev)
2023-12-12T16:52:48.576 INFO:tasks.ceph.osd.0.smithi141.stderr: 1: (ceph::__ceph_assert_fail(char const*, char const*, int, char const*)+0x11e) [0x55d7f04c58ec]
2023-12-12T16:52:48.576 INFO:tasks.ceph.osd.0.smithi141.stderr: 2: ceph-osd(+0x470aa8) [0x55d7f04c5aa8]
2023-12-12T16:52:48.576 INFO:tasks.ceph.osd.0.smithi141.stderr: 3: ceph-osd(+0x4a089f) [0x55d7f04f589f]
2023-12-12T16:52:48.576 INFO:tasks.ceph.osd.0.smithi141.stderr: 4: (Objecter::handle_osd_map(MOSDMap*)+0x18a5) [0x55d7f08eef05]
2023-12-12T16:52:48.577 INFO:tasks.ceph.osd.0.smithi141.stderr: 5: (OSD::handle_osd_map(MOSDMap*)+0x864) [0x55d7f0624294]
2023-12-12T16:52:48.577 INFO:tasks.ceph.osd.0.smithi141.stderr: 6: (OSD::ms_dispatch(Message*)+0x68) [0x55d7f0627e58]
2023-12-12T16:52:48.577 INFO:tasks.ceph.osd.0.smithi141.stderr: 7: ceph-osd(+0x57663b) [0x55d7f05cb63b]
2023-12-12T16:52:48.577 INFO:tasks.ceph.osd.0.smithi141.stderr: 8: (DispatchQueue::entry()+0x5e2) [0x55d7f0dc1b42]
2023-12-12T16:52:48.577 INFO:tasks.ceph.osd.0.smithi141.stderr: 9: ceph-osd(+0xc0bf41) [0x55d7f0c60f41]
2023-12-12T16:52:48.577 INFO:tasks.ceph.osd.0.smithi141.stderr: 10: /lib64/libc.so.6(+0x9f802) [0x7fea8649f802]
2023-12-12T16:52:48.577 INFO:tasks.ceph.osd.0.smithi141.stderr: 11: /lib64/libc.so.6(+0x3f450) [0x7fea8643f450]
```

See https://pulpito.ceph.com/yuriw-2023-12-12_16:18:21-rados-wip-yuri-testing-2023-12-11-1524-distro-default-smithi/ for more examples.

**Comment by Matan-B** [comment]
> @Matan-B there is a crash showing up in teuthology tests that seems related:
 
@ljflores, fixed and rebased -  thank you!

**Comment by ljflores** [comment]
Rados approved: https://tracker.ceph.com/projects/rados/wiki/MAIN#httpstrellocomclTLi41v61905-wip-yuri-testing-2023-12-19-1112-old-wip-yuri-testing-2023-12-13-1239-old-wip-yuri-testing-2023-12-11-1524
